src/bdc/steps/step.py

```diff
@@ -0,0 +1,68 @@
+# SPDX-License-Identifier: MIT
+# SPDX-FileCopyrightText: 2023 Lucca Baumgärtner <lucca.baumgaertner@fau.de>
+
+
+class Step:
+    def __init__(self):
+        self._df = None
+        self._input_location = None
+        self._output_location = None
```

Comment:
This would be a possible class for pydantic. Essentially you define all fields, their types, their value ranges, their default values, etc. during the class definition. pydantic then ensures that every instance of that object adheres to these rules on creation. For this you would do something like this
```
from pydantic import BaseModel
import pandas
from typing import Optional

class Step(BaseModel):
    df: pandas.DataFrame
    input_location: str
    output_location: Optional[str]
```
Here you can see that the type hints from the typing module are used. Any value that is not marked as OPtional will be required when initializing this object. Dependent on what they are you could also make the input and output locations pathlib.Path objects as these offer some intuitive file path handling. 
Generally I added the pydantic module to the lead objects as I assume that they will always need to adhere to some general structure. If the same thing applies to the Step class I am not completely sure. But as you can see also in the database.models.LeadValue object it can be helpful sometimes to limit logical ranges of values, and just overall ensure data consistency. Maybe this will give you some insights where you might want to use the pydantic.BaseModel. If you have any specific question lmk on slack.

Reply:
The rest of the class can stay the same. You would only remove the __init__ function as the BaseModel takes care of that. Note that every data field needs to be declared with key word arguments when using pydantic i.e.
```
step = Step(df=pandas.DataFrame(), input_location="data/file", output_location="data/other") # valid
step = Step(df=pandas.DataFrame(), input_location="data/file) # valid where the Optional field becomes None
step = Step(df=pandas.DataFrame()) # invalid as at least one required field is not set
step = Step(pandas.DataFrame(), "data/file") # invalid as positional arguments are not allowed
```

Reply:
You could also set default arguments so lets say the df should be empty on creation 
```
from pydantic import BaseModel
import pandas
from typing import Optional

class Step(BaseModel):
    df: pandas.DataFrame = pandas.DataFrame()
    input_location: str
    output_location: Optional[str]
```

Reply:
Thanks alot for the comprehensive reply!

Reply:
I added basic typing to the methods. As the `Step()` constructor receives no arguments, I'll leave it at that for now.

Reply:
Sounds good